lang/java/reef-bridge-java/src/main/java/org/apache/reef/bridge/Network.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.reef.wake.remote.ports.TcpPortProvider;
+import org.apache.reef.wake.impl.LoggingEventHandler;
+import org.apache.reef.wake.remote.address.LocalAddressProvider;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The CLR Bridge Network class aggregates a RemoteManager and
+ * Protocol Serializer to provide a simple send/receive interface
+ * between the Java and CLR sides of the bridge.
+ */
+public final class Network {
+  private static final Logger LOG = Logger.getLogger(Network.class.getName());
+  private static final AtomicLong IDENTIFIER_SOURCE = new AtomicLong(0);
+
+  private RemoteManager remoteManager;
+  private InetSocketAddress inetSocketAddress;
+  private EventHandler<byte[]> sender;
+  private final LocalObserver localObserver;
+  private final ProtocolSerializer serializer = new ProtocolSerializer("org.apache.reef.bridge.message");
+
+  /**
+   * Sends and receives messages between the java bridge and C# bridge.
+   * @param localAddressProvider Local address provider used to find open port.
+   * @param observer A multiobserver instance that will receive all incoming messages.
+   */
+  public Network(final LocalAddressProvider localAddressProvider, final MultiObserver observer) {
+    LOG.log(Level.INFO, "Initializing");
+
+    this.localObserver = new LocalObserver(this, observer);
+
+    try {
+      String name = "JavaBridgeNetwork";
+      int port = 0;
+      boolean order = true;
+      int retries = 3;
+      int timeOut = 10000;
+
+      // Instantiate a port provider.
+      final Injector injector = Tang.Factory.getTang().newInjector();
+      final TcpPortProvider tcpPortProvider = Tang.Factory.getTang().newInjector().getInstance(TcpPortProvider.class);
+
+      // Instantiate a remote manager to handle java-C# bridge communication.
+      final RemoteManagerFactory remoteManagerFactory = injector.getInstance(RemoteManagerFactory.class);
+      remoteManager = remoteManagerFactory.getInstance(
+        name, localAddressProvider.getLocalAddress(), port, new ByteCodec(),
+        new LoggingEventHandler<Throwable>(), order, retries, timeOut,
+        localAddressProvider, tcpPortProvider);
+
+      // Get our address and port number.
+      final RemoteIdentifier remoteIdentifier = remoteManager.getMyIdentifier();
+      if (remoteIdentifier instanceof SocketRemoteIdentifier) {
+        SocketRemoteIdentifier socketIdentifier = (SocketRemoteIdentifier)remoteIdentifier;
+        inetSocketAddress = socketIdentifier.getSocketAddress();
+      } else {
+        throw new RuntimeException("Identifier is not a SocketRemoteIdentifier");
+      }
+
+      // Register as the message handler for any incoming messages.
+      remoteManager.registerHandler(byte[].class, this.localObserver);
+
+    } catch (final Exception e) {
+      e.printStackTrace();
```

Comment:
never do that. the logger will write stack trace for you

Reply:
Fixed.